hail/src/main/scala/is/hail/annotations/RegionValueBuilder.scala

```diff
@@ -167,6 +167,33 @@ class RegionValueBuilder(var region: Region) {
     val length = t.loadLength(region, aoff)
     assert(length == indexstk.top)
 
+    endArrayUnchecked()
+  }
+
+  // using this function, rather than startArray will set all elements of the array to missing by
+  // default, you will need to use setPresent to add a value to this array.
+  def startMissingArray(length: Int, init: Boolean = true) {
+    val t = currentType().asInstanceOf[PArray]
+    if (t.elementType.required)
+      fatal(s"cannot use random array pattern for required type ${ t.elementType }")
+    val aoff = region.allocate(t.contentsAlignment, t.contentsByteSize(length))
+
+    if (typestk.nonEmpty) {
+      val off = currentOffset()
+      region.storeAddress(off, aoff)
+    } else
+      start = aoff
+
+    typestk.push(t)
+    elementsOffsetstk.push(aoff + t.elementsOffset(length))
+    indexstk.push(0)
+    offsetstk.push(aoff)
+
+    if (init)
+      t.initialize(region, aoff, length, setMissing = true)
+  }
+
+  def endArrayUnchecked() {
```

Comment:
Can you move this next to `endArray`? Also, your current code may be better for performance reasons, but I would have had one method with a flag `check=true` and then only assert if check is true.